gateway/gateway.go

```diff
@@ -1479,6 +1471,39 @@ func (gateway *HandleT) StartAdminHandler(ctx context.Context) error {
 	return g.Wait()
 }
 
+func (gateway *HandleT) maxActiveClientMiddleware(next http.Handler) http.Handler {
+
+	//this is to make sure that we don't have more than `maxClient` in-memory at any point of time. As, having more http client than `maxClient`
+	// may lead to gateway OOM kill.
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt32(&gateway.activeClientCount, 1)
+		defer func() {
+			atomic.AddInt32(&gateway.activeClientCount, -1)
+		}()
+		if maxActiveClients != 0 {
+			select {
+			case gateway.activeClient <- struct{}{}:
+				defer func() {
+					<-gateway.activeClient
+				}()
```

Comment:
I would separate counting and limit into a separate middleware.
I would also make them self contain and independent from the gateway component, not for reusability but to avoid accidental and unnecessary coupling.

sudo code:
```go
package middleware
func  LimitConcurrentRequests(maxRequest int) func(http.Handler) http.Handler {
    activeClient := make(chan struct{}, maxRequest)
    return func(next.http.Handler) http.Handler {
          ....
          next.ServeHTTP(w, r)
     }


     }
}
````

Reply:
This will make it easy to test as well, please include tests

Reply:
modified as per your feedback.